masonry/src/widget/progress_bar.rs

```diff
@@ -0,0 +1,235 @@
+        );
+        let our_size = bc.constrain(desired_size);
+        trace!("Computed layout: size={}", our_size);
+        our_size
+    }
+
+    fn paint(&mut self, ctx: &mut PaintCtx, scene: &mut Scene) {
+        let border_width = 1.;
+
+        if self.label.needs_rebuild() {
+            debug_panic!("Called ProgressBar paint before layout");
+        }
+
+        let rect = ctx
+            .size()
+            .to_rect()
+            .inset(-border_width / 2.)
+            .to_rounded_rect(2.);
+
+        fill_lin_gradient(
+            scene,
+            &rect,
+            [theme::BACKGROUND_LIGHT, theme::BACKGROUND_DARK],
+            UnitPoint::TOP,
+            UnitPoint::BOTTOM,
+        );
+
+        stroke(scene, &rect, theme::BORDER_DARK, border_width);
+
+        let progress_rect_size = Size::new(
+            ctx.size().width * self.part_complete.unwrap_or(1.) as f64,
+            ctx.size().height,
+        );
+        let progress_rect = progress_rect_size
+            .to_rect()
+            .inset(-border_width / 2.)
+            .to_rounded_rect(2.);
+
+        fill_lin_gradient(
+            scene,
+            &progress_rect,
+            [theme::PRIMARY_LIGHT, theme::PRIMARY_DARK],
+            UnitPoint::TOP,
+            UnitPoint::BOTTOM,
+        );
+        stroke(scene, &progress_rect, theme::BORDER_DARK, border_width);
+
+        // center text
+        let widget_size = ctx.size();
+        let label_size = self.label.size();
+        let text_pos = Point::new(
+            ((widget_size.width - label_size.width) * 0.5).max(0.),
+            ((widget_size.height - label_size.height) * 0.5).max(0.),
+        );
+        self.label.draw(scene, text_pos);
+    }
+
+    fn accessibility_role(&self) -> Role {
+        Role::ProgressIndicator
+    }
+
+    fn accessibility(&mut self, ctx: &mut AccessCtx) {
+        ctx.current_node().set_value(self.value_accessibility())
+    }
+
+    fn children_ids(&self) -> SmallVec<[WidgetId; 16]> {
+        smallvec![]
+    }
+
+    fn make_trace_span(&self) -> Span {
+        trace_span!("Progress bar")
+    }
+
+    fn get_debug_text(&self) -> Option<String> {
+        Some(self.value_accessibility().into())
+    }
+}
+
+fn clamp_part_complete(part_complete: &mut Option<f32>) {
+    if let Some(value) = part_complete {
+        if value.is_nan() {
+            *part_complete = None;
+        } else {
+            *part_complete = Some(value.clamp(0., 1.));
+        }
+    }
+}
+
+// --- MARK: TESTS ---
+#[cfg(test)]
+mod tests {
+    use insta::assert_debug_snapshot;
+
+    use super::*;
+    use crate::assert_render_snapshot;
+    use crate::testing::{widget_ids, TestHarness, TestWidgetExt};
+
+    #[test]
+    fn empty_progressbar() {
+        let [checkbox_id] = widget_ids();
+        let widget = ProgressBar::new(Some(0.)).with_id(checkbox_id);
+
+        let mut harness = TestHarness::create(widget);
+
+        assert_debug_snapshot!(harness.root_widget());
```

Comment:
again, not sure if I've done this right.

Reply:
Note that you've added some snapshot/screenshots tests, but you haven't actually added the snapshots and screenshots yet, so the tests will necessary fail.

Also, you can probably add a few more screenshot tests for different levels of completion in the progress bar. (I'd say 0%, 50%, 100% and None are the ones you're interested in.)